Src/FluentAssertions.Net40/Xml/XmlReaderValidator.cs

```diff
@@ -0,0 +1,239 @@
+        }
+
+        Stack<string> locationStack = new Stack<string>();
+
+        public XmlReaderValidator(XmlReader subjectReader, XmlReader expectedReader, string because, object[] reasonArgs)
+        {
+            assertion = Execute.Assertion.BecauseOf(because, reasonArgs);
+        
+            this.subjectReader = subjectReader;
+            this.expectedReader = expectedReader;
+        }
+                
+        class ValidationResult
+        {
+            public ValidationResult(string formatString, params object[] formatParams)
+            {
+                FormatString = formatString;
+                FormatParams = formatParams;
+            }
+
+            public string FormatString { get; }
+
+            public object[] FormatParams { get; }
+        }
+
+        public void Validate(bool expectedEquivalence)
+        {
+            var validationResult = Validate();
+
+            if(expectedEquivalence && validationResult != null)
+            {
+                assertion.FailWith(validationResult.FormatString, validationResult.FormatParams);
+            }
+            if(!expectedEquivalence && validationResult == null)
+            {
+                assertion.FailWith("Expected Xml to not be equivalent{reason}, but it is.");
+            }
+        }
+
+        private ValidationResult Validate()
+        {
+            subjectReader.MoveToContent();
+            expectedReader.MoveToContent();
+            while (!subjectReader.EOF && !expectedReader.EOF)
+            {
+                if (subjectReader.NodeType != expectedReader.NodeType)
+                {
+                    return new ValidationResult("Expected node of type {0} at {1}{reason}, but found {2}.",
+                        expectedReader.NodeType, CurrentLocation, subjectReader.NodeType);
+                }
+
+                ValidationResult validationResult = null;
+
+                switch (subjectReader.NodeType)
+                {
+                    case XmlNodeType.Element:
+                        validationResult = ValidateStartElement();
+                        if(validationResult != null)
+                        {
+                            return validationResult;
+                        }
+                        locationStack.Push(subjectReader.LocalName);
+                        validationResult = ValidateAttributes();
+                        break;
+                    case XmlNodeType.EndElement:
+                        // No need to verify end element, if it doesn't match
+                        // the start element it isn't valid XML, so the parser
+                        // would handle that.
+                        locationStack.Pop();
+                        break;
+                    case XmlNodeType.Text:
+                        validationResult = ValidateText();
+                        break;
+                    default:
+                        throw new NotImplementedException();
```

Comment:
Other node types are those derived from XmlNode. Things that I think might occur in a document are comments (can just be ignored) and CData. Less likely are entities and notations that I think are DTD related. Comments should just be ignored and CDAta would be good to add support for.

And of course, the NotImplementedException should include the node type.